### PR TITLE
Casted unicode object to string to fix subscription_id string error

### LINKED ISSUE #1322: "Parameter 'subscription_id' must be str." when using SDK Auth File for Client
Recently SDK Auth File support was introduced in PR #1297, but when trying to use it I get the following error: `Parameter 'subscription_id' must be str.` I have seen this before when attempting to authorize a client with a subscription ID from a unicode object read-in from a JSON file, and to fix that error I converted the unicode object to a string.

Now that there is JSON parsing of the new Auth File format in the SDK itself, the unicode object is passed directly into the client instantiation without converting to a string, causing the error. This is only an issue with older versions of Python whose JSON library reads in strings as unicode objects, and should not be an issue with versions newer than Python 3.0.

I am working on a pull request to fix this issue now.

### PR DESCRIPTION
**Pull request by jayden-at-arista**
Fixes #1322

### CODE REVIEW

**Comment by jayden-at-arista** [comment]
I'm aware of the unicode issue failing the travis build, I'll update the PR once it's fixed.

**Comment by lmazuel** [comment]
Hi @jayden-at-arista 
Thanks for the contribution!

May I ask you to add a regression test here:
https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt/tests/test_common.py

Meaning, a test that only succeeds in your code is in.

But the idea of the code LGTM. I might use `six` instead to avoid the python version test. I can't look at it today, but it's high priority for this week for sure.

Thanks!!!

**Comment by lmazuel** [comment]
@brettcannon for advice

File: azure-common/azure/common/client_factory.py [hunk 1]
```diff
@@ -25,6 +26,8 @@ def _instantiate_client(client_class, **kwargs):
     args = get_arg_spec(client_class.__init__).args
     if 'subscription_id' not in args:
         del kwargs['subscription_id']
+    elif sys.version_info < (3, 0) and isinstance( kwargs['subscription_id'], unicode ):
+        kwargs['subscription_id'] = str( kwargs['subscription_id'] )
```
Comment (brettcannon) [inline azure-common/azure/common/client_factory.py]: If you're going to be casting text down to `str` from `unicode` in Python 2.7 then you should encode with an explicit encoding. There's also unnecessary spaces within the parentheses.

**Comment by jayden-at-arista** [comment]
@brettcannon I removed the padding spaces and added explicit encoding. @lmazuel I added regression test, and confirmed it fails without my change and passes with my change. The issue was that the mock clients didn't mock the parameter checking that the real clients perform.

**Comment by jayden-at-arista** [comment]
@lmazuel I feel the code is more self documenting with the explicit version check, as this is a band-aid fix of a bigger incompatibility issue. If we wanted to change this to use the six library, we would (from what I saw when experimenting with it) just check to see if it was of type `six.string_types` and then blindly cast it to `str`, which doesn't really show why the cast is necessary.

**Comment by lmazuel** [comment]
@jayden-at-arista I figured the mock as well at 2am in the night, waking up "damn, the mock" (true story lol). LGTM. @brettcannon have you still a concern?

**Comment by brettcannon** [comment]
@lmazuel nope, single concern was taken care of by @jayden-at-arista 😄

**Comment by lmazuel** [comment]
So we're goot to go! Thanks @jayden-at-arista for the PR!!

**Comment by lmazuel** [comment]
https://pypi.python.org/pypi/azure-common/1.1.8

**Pull request merged**